Add an example for hyphenate-limit-chars (#32199)
mdn/content: files/en-us/web/css/hyphenate-limit-chars/index.md

- Minimum number of characters before the hyphen: 2
- Minimum number of characters after the hyphen: 2

> Note that if a word is too short to meet the given constraints, it will not be hyphenated. For example, given a value like `hyphenate-limit-chars auto 3 4`, words shorter than 7 characters will never be hyphenated, since it is impossible to have 3 characters before the hyphen and 4 characters after it.

Reviewer: ```suggestion
Note that if a word is too short to meet the given constraints, it will not be hyphenated. For example, given a value like `hyphenate-limit-chars: auto 3 4`, words shorter than 7 characters will never be hyphenated, since it is impossible to have 3 characters before the hyphen and 4 characters after it.
```

## Examples

### Setting hyphenation limits

In this example we have four paragraphs each containing the same text, and will set different values for `hyphenate-limit-chars` to see the effect.

#### HTML

```html
<div class="container">
  <p id="ex1">juxtaposition and acknowledgement</p>
  <p id="ex2">juxtaposition and acknowledgement</p>
  <p id="ex3">juxtaposition and acknowledgement</p>

>   <p id="ex4">juxtaposition and acknowledgement</p>

Reviewer: Out of scope for this PR but if you're inclined, we could add something around this to the page:
(from spec)
> If the word does not meet the required minimum number of characters in the word / before the hyphen / after the hyphen, then the word must not be hyphenated. Nonspacing combining marks (Unicode General Category Mn) and intra-word punctuation (Unicode General Category P*) do not count towards the minimum.

The first sentence can come handy to explain if you use `12 7 8` for the last box

</div>
```

#### CSS

```css
.container {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(160px, 1fr));
}

p {
  margin: 1rem;
  width: 120px;
  border: 2px dashed #999;
  font-size: 1.5rem;
  hyphens: auto;
}

#ex2 {
  hyphenate-limit-chars: 14;
}

#ex3 {
  hyphenate-limit-chars: 13 9 3;
}

#ex4 {
  hyphenate-limit-chars: 12 7 5;
}
```

#### Result

{{EmbedLiveSample("Setting hyphenation limits", "", 200)}}

In the first example, we don't set `hyphenate-limit-chars`, and the user agent applies its default algorithm.

In the second example, we specify `hyphenate-limit-chars: 14` meaning that words must be at least 14 characters long before they are hyphenated. In the example, "juxtaposition" is no longer hyphenated, since it is only 13 characters long.

> In the third example, we specify `hyphenate-limit-chars: 13 9 3`, so each hyphenated word must have at least 9 characters before the hyphen. The effect is that "acknowledgement" is now hyphenated as "acknowledge-ment" rather than "acknowl-edgement".

Author: > * We can also explain `13` and `3` here: "juxtaposition" is 13 chars so hyphen gets added after `9` chars. The reaming part after hyphen has more than `3` chars.

Perhaps it isn't clear, but the point is I'm comparing the _difference_ between each of cases 2-4 with the default case 1. That's why I'm not talking about 13 and 3 here. I just want to focus on the places the given values are changing things.

So the idea is that there are four cases presented:
- specify nothing
- specify minimum word length
- specify minimum char length before hyphen
- specify minimum char length after hyphen

...that is, each case presents one aspect only of the property.

It might be better to use the default values (`5 2 2`) in places we're not trying to show a different value, so for instance in example 3 have `5 9 2` rather than `13 9 3`, and that might help signal to the reader that we are here trying to show the effect of specifying a minimum

Also perhaps we could reword to emphasise this, like:

> The second example shows how to prevent words under a certain character length from being hyphenated. We specify hyphenate-limit-chars: 14 meaning that words must be at least 14 characters long before they are hyphenated. In the example, "juxtaposition" is no longer hyphenated, since it is only 13 characters long.

?

> 
>     * For "acknowledgement", should the hyphen get added after "acknowled"? "acknowledge-ment" has 11 chars before the hyphen. Is it because the user agent is applying its algorithm to add the hyphen at the point that satisfies both 9 chars before and 3 chars after and the split parts are still sort of readable?
> 

Yes, exactly. It only promises "at least 9", it's allowed to have more, and it has its own algorithm to determine what's best. That's what makes this a bit tricky to have examples for, because changing one value can change results in slightly unexpected ways.

>     * I don't know how it would be ""acknowl-edgement", maybe we explain this?

As above, that's the default behavior. I don't think we can really explain the default behavior, because that's (AFAIK) up to the browser to implement and is probably quite complicated. Again, though, the point here is only to compare the things that have changed versus the default.

> In the fourth example, we specify `hyphenate-limit-chars: 12 7 5`, so each hyphenated word must have at least 5 characters after the hyphen. The effect is that "juxtaposition" is hyphenated as "juxtapo-sition" rather than "juxtaposi-tion".

Reviewer: - I understand the hyphenation here: both "juxtaposition" and "acknowledgement" are at least 12 chars long, so hyphen is added after 7 chars. So it's "juxtapo-sition" and "acknowl-edgement".
- > so each hyphenated word must have at least 5 characters after the hyphen.

I think we should explain both `7` and `5`.
- Will it help if we show the last value to be `8` instead of `5`? In that case, the chars after hyphen take precedence over chars before hyphen and "juxtaposition" will not get hyphenated. "acknowledgement" will still have 8 chars after hyphen, so will exactly match 7 chars before and 8 chars after the hyphen.
- In this case too, I am not sure why it would be ""juxtaposi-tion".
- Would be nice to explain how `hyphenate-limit-chars` is getting applied to both "juxtaposition" and "acknowledgement".

Author: > * I understand the hyphenation here: both "juxtaposition" and "acknowledgement" are at least 12 chars long, so hyphen is added after 7 chars. So it's "juxtapo-sition" and "acknowl-edgement".

I'm no expert here but this is not my reading of it. `12 7 5` doesn't say that the hyphen must be added after 7 chars. It says that:
- only hyphenate words at least 12 chars long
- there must be _at least_ 7 chars before the hyphen
- there must be _at least_ 5 chars after it.

Beyond that, the browser can do what it likes. So while "juxtapo-sition" is allowed, so is ""juxtapos-ition". Similarly it could do "acknowledg-ement".

We'd see the same results with `5 2 5`, maybe that would be a clearer example?

 
>     * > so each hyphenated word must have at least 5 characters after the hyphen.
> 
> 
> I think we should explain both `7` and `5`.
> 
>     * Will it help if we show the last value to be `8` instead of `5`? In that case, the chars after hyphen take precedence over chars before hyphen and "juxtaposition" will not get hyphenated. 

I'm not sure it is about precedence exactly, more that "juxtaposition" is only 13 characters, so it's not possible to have 7 characters before the hyphen _and_ 8 characters after it.

> "acknowledgement" will still have 8 chars after hyphen, so will exactly match 7 chars before and 8 chars after the hyphen.

Indeed, because "acknowledgement" is 15 characters, then yes, `13 7 8` exactly  specifies how it must break.

> 
>     * In this case too, I am not sure why it would be ""juxtaposi-tion".

Again, just because that's what the algorithm would choose, if we hadn't constrained it to have at least 5 characters after the hyphen.
> 
>     * Would be nice to explain how `hyphenate-limit-chars` is getting applied to both "juxtaposition" and "acknowledgement".

Reviewer: Ha! When I wrote "I understand the hyphenation here: both "juxtaposition" and "acknowledgement" are at least 12 chars long, so hyphen is added after 7 chars. So it's "juxtapo-sition" and "acknowl-edgement", I actually did mean there should be 7 chars _before_ the hyphen (so then hyphen is added after 7 chars).

> - only hyphenate words at least 12 chars long
> - there must be at least 7 chars before the hyphen
> - there must be at least 5 chars after it.

Agree! That's how I see this property working as well.

> I'm not sure it is about precedence exactly, more that "juxtaposition" is only 13 characters, so it's not possible to have 7 characters before the hyphen and 8 characters after it.

Which is why it might be a good thing to add to prose a rephrased version of this statement from the spec (https://github.com/mdn/content/pull/32199#discussion_r1486585214):

**_If the word does not meet the required minimum number of characters in the word / before the hyphen / after the hyphen, then the word must not be hyphenated._**

Can we add this to the [`x y z`](https://developer.mozilla.org/en-US/docs/Web/CSS/hyphenate-limit-chars#x_y_z) value description?

Author: Yes, I think we should. Also worth an example, not a proper full one but just something like "For example, given value like `5 4 3`, words less than 7 characters long will never be hyphenated, because it it impossible to have 4 characters before the hyphen and 3 characters after it.

I think it might be better having this at the end of "Syntax" , though, as it's not only applicable to `x y z`.

Author: https://github.com/mdn/content/pull/32199/commits/cb4f636544209f54f3510695a846d9655ab16bdb adds a paragraph on this.

#### HTML

```html
<div class="container">
  <p id="ex1">juxtaposition and acknowledgement</p>
  <p id="ex2">juxtaposition and acknowledgement</p>
  <p id="ex3">juxtaposition and acknowledgement</p>
  <p id="ex4">juxtaposition and acknowledgement</p>
</div>
```

#### CSS

```css
.container {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(160px, 1fr));
}

p {
  margin: 1rem;
  width: 120px;
  border: 2px dashed #999;
  font-size: 1.5rem;
  hyphens: auto;
}

#ex2 {
  hyphenate-limit-chars: 14;
}

#ex3 {
  hyphenate-limit-chars: 5 9 2;
}

#ex4 {
  hyphenate-limit-chars: 5 2 7;
}
```

#### Result

{{EmbedLiveSample("Setting hyphenation limits", "", 200)}}

In the first box, we don't set `hyphenate-limit-chars`, allowing the browser to apply its default algorithm. By default, the browser uses the values `5 2 2` unless it can find better values.

In the second box, we prevent the browser from hyphenating words unless they are at least 14 characters long by setting `hyphenate-limit-chars: 14`. As a result, "juxtaposition" is not hyphenated in the second box because it is only 13 characters long.

> In the third box, we will make the browser include at least 9 characters before the hyphen. We do this by setting `hyphenate-limit-chars: 5 9 2`. The effect is that "acknowledgement" is now hyphenated as "acknowledge-ment" rather than "acknowl-edgement".

Author: ```suggestion
In the third box, we constrain the browser to include at least 9 characters before the hyphen by setting `hyphenate-limit-chars: 5 9 2`. The effect is that "acknowledgement" is now hyphenated as "acknowledge-ment" rather than the default version "acknowl-edgement", as shown in the first box.

Note that the browser does not have to include exactly 9 characters before the hyphen: as long as the constraints given in `hyphenate-limit-chars` are satisfied, the browser can break the word in the place it considers best. So in this case, for example, it chooses "acknowledge-ment" rather than the less readable "acknowled-gement".
```

Does this seem OK to you?

Reviewer: Yes, this is good, thanks!

## Examples

### Setting hyphenation limits

> In this example, we have four boxes each containing the same text. We'll set different values for `hyphenate-limit-chars` to see the effect.

Reviewer: The first two sentences in "Result" can be moved here to completely describe the example scenario. Something like this (feel free to rephrase):
```suggestion
In this example, we have four boxes each containing the same text. For the purpose of comparison, the first box shows the default hyphenation applied by the browser. The next three boxes demonstrate the result of constraining the browser's default behavior using different `hyphenate-limit-chars` values.
```

#### HTML

```html
<div class="container">
  <p id="ex1">juxtaposition and acknowledgement</p>
  <p id="ex2">juxtaposition and acknowledgement</p>
  <p id="ex3">juxtaposition and acknowledgement</p>
  <p id="ex4">juxtaposition and acknowledgement</p>
</div>
```

#### CSS

```css
.container {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(160px, 1fr));
}

p {
  margin: 1rem;
  width: 120px;
  border: 2px dashed #999;
  font-size: 1.5rem;
  hyphens: auto;
}

#ex2 {
  hyphenate-limit-chars: 14;
}

#ex3 {
  hyphenate-limit-chars: 5 9 2;
}

#ex4 {
  hyphenate-limit-chars: 5 2 7;
}
```

#### Result

> This example illustrates different constraints that `hyphenate-limit-chars` can apply to the browser's hyphenation algorithm.
> 
> We'll first show the default hyphenation applied by the browser, and then see the effect of constraining the default behavior using the `hyphenate-limit-chars` property.
> 

Reviewer: Can be combined with the example description on line 71.
```suggestion
```

{{EmbedLiveSample("Setting hyphenation limits", "", 200)}}

> In the first box, we don't set `hyphenate-limit-chars`, allowing the user agent to apply its default algorithm. This will use values of `5 2 2`, unless the browser can find better values.

Reviewer: - For consistency, could you also change "user agent" to "browser" on [line 51](https://github.com/mdn/content/blob/main/files/en-us/web/css/hyphenate-limit-chars/index.md?plain=1#L51)
- Made the second sentence a bit more direct
```suggestion
In the first box, we don't set `hyphenate-limit-chars`, allowing the browser to apply its default algorithm. By default, the browser uses the values `5 2 2` unless it can find better values.
```

> In the second box, we will prevent the browser hyphenating words unless they are at least 14 charcters long. We do this by setting `hyphenate-limit-chars: 14`. The effect is "juxtaposition" is no longer hyphenated, since it is only 13 characters long.

Reviewer: - Removing future tense plus some other tweaks for clarity.
- Repeating "in the second box" to emphasize on comparison with the first box.
```suggestion
In the second box, we prevent the browser from hyphenating words unless they are at least 14 characters long by setting `hyphenate-limit-chars: 14`. As a result, "juxtaposition" is not hyphenated in the second box because it is only 13 characters long.
```

> In the third box, we will make the browser include at least 9 characters before the hyphen. We do this by setting `hyphenate-limit-chars: 5 9 2`. The effect is that "acknowledgement" is now hyphenated as "acknowledge-ment" rather than "acknowl-edgement".

Reviewer: - It's the last part that was tripping me. I wasn't able to get "rather than "acknowl-edgement"". Of course, I was to look at the first box for comparison for the default behavior. Not sure if someone else has the same issue, but I've tried to clarify this in the last bit of the sentence.
- Should we also mention that, after satisfying the constraints, the browser can further apply its algorithm to place the hyphen? I've added the last sentence. It is verbose but I hope it clarifies why the hyphen is not added after the "d", resulting in 9 chars before the hyphen.
```suggestion
In the third box, we constrain the browser to include at least 9 characters before the hyphen by setting `hyphenate-limit-chars: 5 9 2`. The effect is that "acknowledgement" is now hyphenated as "acknowledge-ment" rather than the browser's default "acknowl-edgement", as shown in the first box. For "acknowledge-ment", the browser adheres to the requirement that there be at least 9 characters before the hyphen and at least 2 characters after it, then applies its own algorithm to determine the hyphen's placement.
```

> In the fourth box, we will make the browser include at least 7 characters after the hyphen. We do this by setting
> `hyphenate-limit-chars: 5 2 7`. The effect is that "juxtaposition" is hyphenated as "juxta-position" rather than "juxtaposi-tion".

Reviewer: This can be minimal considering we've done much of the explanation for third box.
```suggestion
In the fourth box, we make the browser include at least 7 characters after the hyphen by setting
`hyphenate-limit-chars: 5 2 7`. The effect is that "juxtaposition" is hyphenated as "juxta-position" rather than the default "juxtaposi-tion".
```
